### DISCUSSION TITLE
Migrate Protractor automation-UI tests on NWjs to WendriverIO

### DISCUSSION THREAD

**Discussion by YehielShamir** [webdriverio/codemod#23] [Migration Issue]
Hi, 
I have an issue with migration from protractor to webdriverIO 
Currently, I am automating a node-WebKit application (based on chromium) with Protractor.
In order to run the tests, there is a chrome driver binary inside the application's directory.
[This is how node-WebKit recommended](https://docs.nwjs.io/en/latest/For%20Users/Advanced/Test%20with%20ChromeDriver/)
Here is the conf.js being used:
```
exports.config = {
     directConnect : true,
     chromedriver : 'path/to/my/app/chromedriver.exe',
     framework : 'jasmine', 
     restartBrowserBetweenTests : false,
     suites : [
         test/specs/**/*.js
     ],
     onComplete() {},
     onPrepare () {}, 
     onCleanUp() {}
} 
```
Now, I want to migrate to webdriverIO (especially after the angular team announced the protractor would no longer be supported at the end of 2022). 
With webdriverIO, even though I found at the Docs, that webdriverIO supports native desktop applications, I was
not able to find any explanations about how to config webdriverIO.
I have tried to adjust the protractor's config file with `@wdio/codemod` to be compatible with WebdriverIO,
but got an error saying "webdriverIO does not support 'chromedirver' option ". and that I should try to install the `wdio-chromedriver-service` package which also did not fix the problem.

So far, any other tool did not support node WebKit. and because of the great similarity between webdriverIO and protractor, 
I believe it should be achievable.

(i.e. I have been struggling to find any other tools to replace my protractor's automation tests on my node WebKit app )

In conclusion, How do I configure webdriverIO to run an automation on a node WebKit application?

**Comment by christian-bromann** [comment]
> but got an error saying "webdriverIO does not support 'chromedirver' option "

Can you paste this error message here?

> and that I should try to install the `wdio-chromedriver-service` package which also did not fix the problem.

You need to install that package and add it to the service list.

> So far, any other tool did not support node WebKit. and because of the great similarity between webdriverIO and protractor,
> I believe it should be achievable.

It is definitely achievable.

> In conclusion, How do I configure webdriverIO to run an automation on a node WebKit application?

I haven't automated NW.js apps myself so bare with me here. It seems they say they have a custom Chromedriver, where can you download that?

**Reply by YehielShamir** [reply]
> Can you paste this error message here?

I am on a close network (NOT the world wide web). so I cannot paste the exact error here.
But I will try to write more or less the error I got when I trying to migrate the Protractor conf..
```
npx jscodeshift -t ./node_modules/@wdio/codemod/protractor ./conf.js
processing 1 files...
...
Error transforming ./conf.js:
>      chromedriver: 'C/my app/chromedriver.exe'
The option "chromedriver" is not supported in WebdriverIO. We advise to use the "wdio-chromedriver-service" instead.
For more information on this configuration, see https://www.npmjs.com/package/wdio-chromedriver-service.
all done.
Results:
1 error
0 unmodified
0 skipped
0 ok
```
> You need to install that package and add it to the service list.

So, as you offer now, I installed the wdio-chromedriver-service package, and then entered in my new wdio.conf.js file the following config :
```
directConnect: true, 
services : ['chromedriver', {
      protocol : 'http',
      hostname : 'localhost', 
      port : 9515 || 4444, 
      path : 'C:/my app/chromedriver.exe'
}],
...
```
The above syntax appeared in @wdio/codemod's README.md file
But still, a got an error as the following:
```
npx wdio run ./wdio.conf.js
Execution of 1 spec files started 
@wdio/utils: initializeServices: Error : Couldn't initialise "wdio-chromedriver-service".
Error : Cannot find module 'chromedriver'
Reqire stack:
... 
INFO @wdio/cli:launcher: Run onPrepare hook
INFO @wdio/cli:launcher: Run onWorkerStart hook
INFO @wdio/local-runner: Start worker 0-0 with arg: run,./wdio.conf.js
INFO @wdio/local-runner: Run worker command: run
ERROR @wdio/local-runner: Error: Type check for option "services" failed: the "services" options need to be a lost of strings
and/or arrays
...
```
The last error is unclear to me . becuase is says that the "services" key supposed to contains list of strings and/or array
and according to @wdio/codemod README file , it is also contains strings and object. 

> I haven't automated NW.js apps myself so bare with me here. It seems they say they have a custom Chromedriver, where can you download that?

as I refer abocve , from the official NW.js website . 
[Here is the docs]()https://docs.nwjs.io/en/latest/For%20Users/Advanced/Test%20with%20ChromeDriver/

**Reply by christian-bromann** [reply]
> The last error is unclear to me .

It should be this way:

```js
directConnect: true, 
services : [
    ['chromedriver', {
        protocol : 'http',
        hostname : 'localhost', 
        port : 9515 || 4444, 
        path : 'C:/my app/chromedriver.exe'
    }],
]
```

I think this solution should work

**Reply by christian-bromann** [reply]
I believe we can add Chromedriver automatically, added an issue for it #24

**Reply by YehielShamir** [reply]
> I believe we can add Chromedriver automatically, added an issue for it #24

I gusset this can be a terrific solution 😊

> It should be this way:

Yes , this was solved my errorType . 
But unfortunately I am still getting the same error above 
```
npx wdio run ./wdio.conf.js
Execution of 1 spec files started 
@wdio/utils: initializeServices: Error : Couldn't initialise "wdio-chromedriver-service".
Error : Cannot find module 'chromedriver'
Reqire stack:
... 

```

**Reply by christian-bromann** [reply]
> Error : Cannot find module 'chromedriver'

Please run:

```sh
$ npm install --save-dev chromedriver
```
